src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -48,77 +57,81 @@ prepLookups ::
-      -- estimate here, and we grow the vector if needed. TODO: tune the
-      -- starting estimate based on the expected true- and false-positives.
-      res <- VUM.unsafeNew (2 * ksN)
-      resixVar <- newPrimVar 0
-      res' <- go res resixVar
-      resix <- readPrimVar resixVar
-      -- The result vector is not guaranteed to be full, so we slice it. This
-      -- means it might retain memory that is not used by the vector.
-      pure $ VUM.slice 0 resix res'
+bloomQueries ::
+     V.Vector (RunLookupView h)
+  -> V.Vector SerialisedKey
+  -> Word32
+  -> VU.Vector (RunIx, KeyIx)
+bloomQueries !rs !ks !resN
+  | rsN == 0 || ksN == 0 = VU.empty
+  | otherwise            = VU.create $ do
+      res <- VUM.unsafeNew (fromIntegral resN)
+      loop res 0 0 (rlvBloom $ rs `V.unsafeIndex` 0) 0
   where
     !rsN = V.length rs
     !ksN = V.length ks
 
-    go :: forall s.
-         VUM.MVector s (RunIx, KeyIx)
-      -> PrimVar s ResIx
+    -- Loop over all run and key indexes
+    loop ::
+          VUM.MVector s (RunIx, KeyIx)
+      -> RunIx
+      -> KeyIx
+      -> Bloom SerialisedKey
+      -> ResIx
       -> ST s (VUM.MVector s (RunIx, KeyIx))
-    go res0 resixVar = loop1 res0 0
-      where
-        -- Loop over all run indexes
-        loop1 ::
-             VUM.MVector s (RunIx, KeyIx)
-          -> RunIx
-          -> ST s (VUM.MVector s (RunIx, KeyIx))
-        loop1 !res1 !rix
-          | rix == rsN = pure res1
-          | otherwise
-          = do
-              res1' <- loop2 res1 0 (rlvBloom $ rs `V.unsafeIndex` rix)
-              loop1 res1' (rix+1)
-          where
-            -- Loop over all key indexes
-            loop2 ::
-                 VUM.MVector s (RunIx, KeyIx)
-              -> KeyIx
-              -> Bloom SerialisedKey
-              -> ST s (VUM.MVector s (RunIx, KeyIx))
-            loop2 !res2 !kix b
-              | kix == ksN = pure res2
-              | let !k = ks `V.unsafeIndex` kix
-              , Bloom.elem k b = do
-                  resix <- readPrimVar resixVar
-                  -- Grows the vector if we've reached the end. TODO: tune how
-                  -- much much we grow the vector each time based on the
-                  -- expected true- and false-positives.
-                  res2' <- if resix == VUM.length res2
-                            then VUM.unsafeGrow res2 ksN
-                            else pure res2
-                  VUM.unsafeWrite res2 resix (rix, kix)
-                  writePrimVar resixVar $! (resix +1)
-                  loop2 res2' (kix+1) b
-              | otherwise = loop2 res2 (kix+1) b
+    loop !res !rix !kix !b !resix
+      -- The result vector is not guaranteed to be full, so we slice it. This
+      -- means it might retain memory that is not used by the vector.
+      | rix == rsN = pure $ VUM.slice 0 resix res
+      | kix == ksN = loop res (rix+1) 0 (rlvBloom $ rs `V.unsafeIndex` rix) resix
```

Comment:
Yeah it's tricky to flatten this one into one loop rather than two nested loops.

Note here that the selection of the next b is wrong: it's picking `rs ! rix` rather than `rs ! rix+1`, but if it does us `rix+1` then it is going to index out of bounds on the last iteration!

So that's why I didn't in the end recommend flattening this one into a single loop (just getting rid of the top level `go`).

Reply:
Do you want me to change it back, or just fix the off-by-one bug?

Reply:
It's not clear to me how to fix the off-by-one bug without introducing a bug where you index outside the bounds of the array of bloom filters, while keeping the flat structure. If you can see a clean way to do that then sure, otherwise I'd suggest reverting to the nested style where that problem doesn't arise.

Reply:
Made it two loops again